### DISCUSSION TITLE
What happens when multiple processes call UseLocalhostClustering() using the same silo port in Orleans?

### DISCUSSION THREAD

**Discussion by alex-vovk** [dotnet/orleans#9124] [Q&A]
I have multiple .NET console applications that use Orleans. 
Each application configures Orleans using 'UseLocalhostClustering()' method without parameters. For each application, the client code runs in the same process as the grain code (aka co-hosted clients).

When I run multiple applications or multiple instances of the same application at the same time, they do not fail, work as expected, and no warnings come from Orleans. However, only one of the running processes listens on ports 11111 and 30000.

My questions:
1. What happens to applications that are not able to listen on a silo port?
2. Why does not Orleans log a warning when it cannot start listening on a default silo port?
3. Does using UseLocalhostClustering() this way in multiple processes have any negative effect on applications?

**Comment by MickaelThumerel** [comment]
UseLocalhostClustering() is only a dev/test mode, you need to mount a DB through Docker or other to connect multiple silo.

**Reply by alex-vovk** [reply]
@MickaelThumerel ,
These applications are independed from each other, they do not need to connect to each other, and clients are co-hosted inside each application.

**Reply by MickaelThumerel** [reply]
@alex-vovk 
'UseLocalhostClustering()' will setup the default endpoint information ip and port. 
That why client end node could connect without any configuration by default.

But in your case you want "MUTLI-CLUSTER" with One Client.

Then you need to specify the endpoint information through ".ConfigureEndpoints()" for each cluster/silo to be sure they use different port.
In the client side, specify multiple cluster ports. "UseLocalhostClustering" in client side could have multiple port to connect.

https://github.com/dotnet/orleans/blob/2d21b3f9afcdd491ded8246ec9ce2401ccbbd388/src/Orleans.Core/Core/ClientBuilderExtensions.cs#L179

**Reply by alex-vovk** [reply]
@MickaelThumerel,
I do not need clients to connect to multiple clusters. So I am not using multiple clusters with one client.

I want each client to connect only to grains hosted in the same process where the client hosted. Each process, in my case, is an independent application that does not need to communicate with other processes that use Orleans.

Ideally, I would like to prevent Orleans runtime from listening on any ports and allow grain calls only within each process. But it looks like this is not supported.

So I am trying to understand the implications of using `UseLocalhostClustering` in multiple independent processes running on the same machine.

**Reply by MickaelThumerel** [reply]
@alex-vovk 
I understand correctly you want to have independent binary with inside a silo ...
No need of client then.

Just create an ApplicationHost to create and host orlean/silo environment and used the grain behaviour inside.

A client is to consume a cluster from outside it.

I'm not sure you can prevent any port listening.
